daft/catalog/__init__.py

```diff
@@ -129,35 +134,63 @@ def register_python_catalog(catalog: PyIcebergCatalog | UnityCatalog, name: str
         >>> daft.catalog.register_python_catalog(catalog, "my_daft_catalog")
 
     """
-    _PYICEBERG_AVAILABLE = False
-    try:
-        from pyiceberg.catalog import Catalog as PyIcebergCatalog
+    if (c := Catalog._try_from(catalog)) is not None:
+        return native_catalog.register_python_catalog(c, name)
+    raise ValueError(f"Unsupported catalog type: {type(catalog)}")
+
+
+class Catalog(ABC):
+    """Interface for python catalog implementations."""
 
-        _PYICEBERG_AVAILABLE = True
-    except ImportError:
-        pass
+    @property
+    def inner(self) -> object | None:
+        """Returns the inner catalog object if this is an adapter."""
+
+    @staticmethod
+    def _try_from(obj: object) -> Catalog | None:
+        for factory in (Catalog._try_from_iceberg, Catalog._try_from_unity):
+            if (c := factory(obj)) is not None:
+                return c
+        return None
 
-    _UNITY_AVAILABLE = False
-    try:
-        from daft.unity_catalog import UnityCatalog
+    @staticmethod
+    def _try_from_iceberg(obj: object) -> Catalog | None:
```

Comment:
I think most accurately, `from_pyiceberg`

Reply:
True. I went for consistency with `daft.read_iceberg`, but could go either way.